File: 024/24.php
<?php

function permute($str, $f = '') {
	$p = [];
	if (strlen($str) == 2) {
		$p[] = $f . $str;
		$p[] = $f . strrev($str);
	} else {
		for ($i = 0; $i < strlen($str); $i++) {
			$first = $str[$i];
			$restStr = substr($str, 0, $i) . substr($str, $i + 1);
			$p = array_merge($p, permute($restStr, $f . $first));
		}
	}
	return $p;
}

$str = "0123456789";
$res = permute($str);

var_dump($res[999999]);

File: 024/24.go
package main

import (
	"fmt"
)

func permute(nums, f []int) [][]int {
	p := [][]int{}
	if len(nums) <= 2 {
		reversedNums := []int{nums[1], nums[0]}
		p = append(p, append(f, nums...))
		p = append(p, append(f, reversedNums...))
	} else {
		for i := 0; i < len(nums); i++ {
			first := nums[i]
			tmp := make([]int, len(nums))
			copy(tmp, nums)
			rest := append(tmp[0:i], tmp[(i+1):]...)

			newP := permute(rest, append(f, first))
			p = append(p, newP...)
		}
	}
	return p
}

func main() {
	nums := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
	res := permute(nums, []int{})
	fmt.Println(res[999999])
}
